### PR TITLE
[eslint-patch] Allow for extended shareable ESLint configurations to use any name

### PR DESCRIPTION
**Pull request by D4N14L**
## Summary

Allows resolution of extended shareable configurations even when the package they are sourced from is not prefixed with `eslint-config-`.

## Details

In order to allow use of ESLint configurations provided via Heft rigs, we need to allow consumption of configurations from packages not prefixed with `eslint-config-`, since Heft rigs obviously do not have the same naming restrictions.

The solution first attempts to resolve using the default logic, and falls back to attempting to resolve using the provided extended config name directly, without modification. This is done to allow for backwards-compatibility with intentionally short-named extended configs (cases where the prefix is intentionally omitted), since ESLint prefixes the `eslint-config-` to the package name during normalization if it was not found. For example, if `@rushstack` was specified as an extends configuration, then it would normally resolve to `@rushstack/eslint-config`. 

## How it was tested

Tested using resolution logic in Rushstack by moving the ESLint configuration into the rig. A future change will transition Rushstack to use the `@rushstack/eslint-config` package via `@rushstack/heft-node-rig` and `@rushstack/heft-web-rig`.

### CODE REVIEW

**Comment by octogonz** [comment]
> In order to allow use of ESLint configurations provided via Heft rigs, we need to allow consumption of configurations from packages not prefixed with eslint-config-, since Heft rigs obviously do not have the same naming restrictions.

This doesn't seem true to me.  We can chat about it offline and then follow up.

File: eslint/eslint-patch/src/_patch-base.ts [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    // For ESLint >=8.0.0, all @eslint/eslintrc code is bundled at this path:
+    //   .../@eslint/eslintrc/dist/eslintrc.cjs
+    try {
+      const eslintrcFolder = path.dirname(
+        require.resolve('@eslint/eslintrc/package.json', { paths: [currentModule.path] })
+      );
+
+      // Make sure we actually resolved the module in our call path
+      // and not some other spurious dependency.
+      if (path.join(eslintrcFolder, 'dist/eslintrc.cjs') === currentModule.filename) {
```
Comment (dmichon-msft) [inline eslint/eslint-patch/src/_patch-base.ts:42]: Wouldn't it be more efficient to verify `currentModule.filename` ends with `dist/eslintrc.cjs` before calling `require.resolve`, then verify that the output of path.dirname matches the remainder of `currentModule.filename`?

File: eslint/eslint-patch/src/_patch-base.ts [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    try {
+      const eslintCandidateFolder = path.dirname(
+        require.resolve('eslint/package.json', {
+          paths: [currentModule.path]
+        })
+      );
+
+      // Make sure we actually resolved the module in our call path
+      // and not some other spurious dependency.
+      if (path.join(eslintCandidateFolder, 'lib/cli-engine/cli-engine.js') === currentModule.filename) {
```
Comment (dmichon-msft) [inline eslint/eslint-patch/src/_patch-base.ts:65]: Some concept here

File: eslint/eslint-patch/src/_patch-base.ts [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+      // For ESLint >=7.8.0, config-array-factory.js is at this path:
+      //   .../@eslint/eslintrc/lib/config-array-factory.js
+      try {
+        const eslintrcFolder = path.dirname(
+          require.resolve('@eslint/eslintrc/package.json', {
+            paths: [currentModule.path]
+          })
+        );
+
+        if (path.join(eslintrcFolder, '/lib/config-array-factory.js') == currentModule.filename) {
```
Comment (dmichon-msft) [inline eslint/eslint-patch/src/_patch-base.ts:98]: Why the switch to `==` from `===`?

Comment (dmichon-msft) [reply eslint/eslint-patch/src/_patch-base.ts:98]: Also, same question

File: eslint/eslint-patch/src/_patch-base.ts [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+          'If you are using a newer ESLint version that may be unsupported, please create a GitHub issue:\n' +
+          'https://github.com/microsoft/rushstack/issues'
+      );
+    }
+    currentModule = currentModule.parent;
+  }
+}
+
+// Detect the ESLint package version
+const eslintPackageJson = fs.readFileSync(path.join(eslintFolder, 'package.json')).toString();
```
Comment (dmichon-msft) [inline eslint/eslint-patch/src/_patch-base.ts:168]: Does calling `toString()` provide some advantage over specifying the encoding in the `readFileSync` call? I'd expect the latter to potentially optimize internally for different buffering behavior.

**Review by dmichon-msft — COMMENTED** [review]
Few questions about the resolution logic, though it looks like you mostly just moved it from another file.

**Comment by iclanton** [comment]
Note: PR #3337 has renamed our GitHub `master` branch to `main`. This should not affect your pull request, but we recommend to redo your `git clone` to avoid confusion with the old branch name.

**Comment by D4N14L** [comment]
Superseded by #4353

**Pull request closed without merging**